rfcs/33-apps_own_scheduling.md

```diff
@@ -0,0 +1,239 @@
+# Feature Name: `apps_own_scheduling`
+
+## Summary
+
+All scheduling information is attached directly to systems, and ultimately owned and configurable by the central app (or hand-written `bevy_ecs` equivalent).
+To ensure that this can be done safely, plugins specify invariants about the ordering of their systems which must be respected.
+Hard ordering constraints (where systems must be both ordered and separated by a hard sync point) are introduced as an assert-style protection to guard against a serious and common form of plugin misconfiguration.
+
+## Motivation
+
+The current plugin model works well in simple cases, but completely breaks down when the central consumer (typically, an `App`) needs to configure when and if their systems run. The main difficulties are:
+
+- 3rd party plugins can have "unfixable" ordering ambiguities or interop issues with other 3rd party plugins
+- plugins can not be added to states, and run criteria cannot be added
+- plugins can not be incorporated into apps and games that use a non-standard control flow architecture
+- end users who care about granular scheduling control for performance optimization have no way to change when systems run
+- worse, plugins can freely insert their own stages, causing sudden and uncontrollable bottlenecks
+- plugins cannot be used by standalone `bevy_ecs` users
+
+See [Bevy #2160](https://github.com/bevyengine/bevy/issues/2160) for more background on this problem.
+
+By moving the ultimate responsibility for scheduling configuration to the central app, users can comfortably combine plugins that were not originally designed to work together correctly and fit the logic into their particular control flow.
+
+## User-facing explanation
+
+**Plugins** are cohesive, drop-in units of functionality that you can add to your Bevy app.
+Plugins can initialize **resources** to store data to, or add **system sets**, used to add coherent batches of logic to your app.
+
+Plugins can be added to your app using the `App::add_plugin` method.
+If you're new to Bevy, third-party plugins added in this way should immediately work without further configuration.
+
+```rust
+use bevy::prelude::*;
+use bevy_hypothetical_tui::prelude::*;
+
+fn main(){
+  App::minimal().add_plugin(TuiPlugin).run();
+}
+```
+
+You can write your own plugins by exporting a function that returns a `Plugin` struct with the desired system sets and resources.
+This can be provided as a standalone function or, more commonly, as a method on a simple `MyPlugin` struct:
+
+```rust
+struct CombatPlugin;
+
+impl IntoPlugin for CombatPlugin {
+ fn into() -> Plugin {
+    Plugin::default()
+      .add_systems("Combat", SystemSet::new().with(attack_system).with(death_system));
```

Comment:
I don't like adding labels like this or "promoting system sets:

1. It creates two different ways to label SystemSets and Systems (the label parameter in the add_systems function and SystemSet::label). It also picks one arbitrary SystemSet label to treat as the "primary key" instead of allowing people to look up the set using any of its labels.
2. This is "pub everything". It forces Plugin authors to expose everything, even if they would like to hide implementation details.
3. It forces literally everything to have a label, which isn't ergonomic.
4. It makes SystemSets "special" instead of being ergonomic ways to insert more than one system with the same configuration. Now users need to think about how a Plugin author decided to insert their systems. Did they use a set? What label did they use as the primary key? This is way too complicated. 

I think we could solve these problems while still allowing "customizable plugins" in the way proposed in this RFC by making the following changes:

1. Remove the label parameter. `add_systems` accepts a single parameter (the system set). Same goes for `add_system`.
2. Remove the single "primary label" as the way to access systems and systems sets.
3. Remove the ability to query for SystemSets. They are back to being ergonomic ways to insert many systems with the same config.
4. (this is the big one) add the ability for an App to query for systems arbitrary labels.

This would look something like this:

```rust 
struct CombatPlugin;

impl IntoPlugin for CombatPlugin {
 fn into() -> Plugin {
    Plugin::default()
      .add_systems(SystemSet::new()
          .label("Combat")
          .with(attack_system)
          .with(death_system)
       );
      .add_system(player_movement_system.label("PlayerMovement")
      .init_resouce::<Score>()
  }
}

fn main(){
  App::default()
    .add_plugin(CombatPlugin)
    .labeled_systems("Combat", |systems| {
      for system in systems {
        system.label("Foo").after("Bar")
      }
    })
    .run();
}
```

This still forces Plugin authors to define "sets" of systems (via labels) _and_ forces them to think about the "sets" they are exporting (by nature of deciding which labels to export). Labels do everything that SystemSets can do. I see no real benefit to promoting SystemSets to be a part of the "public structure" of a bevy app and doing so has significant downsides.

Additionally, if we _really_ need "pub everything", I'd prefer to do that implicitly via "automatic labels". Ex: a SystemName(String) label and a Plugin(TypeId) label. I personally don't like that because its still "pub everything", but if we ultimately agree that we need such a feature, I think the "auto label" approach is better.